woocommerce/woocommerce-gateway-stripe
# Handle UPE free trial subscriptions with SEPA

# Changes proposed in this Pull Request:

During testing it was discovered that the following combination was not being processed correctly:
- UPE enabled
- Free trial subscription
- Using a non-card payment method

When trying to purchase a free trial subscription using UPE and a payment method other than card, a notice was thrown and the order was not processed correctly:

```
Notice: Undefined property: stdClass::$card in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-method-cc.php on line 114

Call Stack:
    0.0055     411360   1. {main}() /var/www/html/index.php:0
    0.0057     411632   2. require('/var/www/html/wp-blog-header.php') /var/www/html/index.php:17
    0.9187    9960856   3. wp() /var/www/html/wp-blog-header.php:16
    0.9187    9960880   4. WP->main() /var/www/html/wp-includes/functions.php:1291
    0.9364   10068456   5. do_action_ref_array() /var/www/html/wp-includes/class-wp.php:763
    0.9364   10068456   6. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:518
    0.9364   10068456   7. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:327
    0.9392   10071104   8. WC_Stripe_Intent_Controller->maybe_process_upe_redirect() /var/www/html/wp-includes/class-wp-hook.php:303
    0.9823   10230080   9. WC_Stripe_UPE_Payment_Gateway->maybe_process_upe_redirect() /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php:673
    0.9832   10230200  10. WC_Stripe_UPE_Payment_Gateway->process_upe_redirect_payment() /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:725
    1.0275   10315408  11. WC_Stripe_UPE_Payment_Gateway->process_order_for_confirmed_intent() /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:752
    1.3855   10405872  12. WC_Stripe_UPE_Payment_Method_CC->add_token_to_user_from_intent() /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:796
    1.3855   10405872  13. WC_Stripe_UPE_Payment_Method_CC->add_token_to_user_from_payment_method() /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-method-cc.php:79
    1.3860   10406344  14. WC_Stripe_UPE_Payment_Method_CC->create_payment_token_for_user() /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-upe-payment-method-cc.php:95
```

This was due to incorrectly getting the payment method data from a setup intent. The previous code was ending up with card as the payment method each time and thus it was unable to save the payment method for future use. This has been updated now to properly retrieve the payment method details on a setup intent and then process the order successfully.

# Testing instructions

- Install [Stripe CLI](https://stripe.com/docs/stripe-cli).
- Run `stripe login` and select the Stripe account you're connected to.
- Run `stripe listen --forward-to 'http://localhost:8082/?wc-api=wc_stripe'` and copy the webhook signing secret (`whsec_******`) to the plugin settings.

Setup:
- This branch pulled down
- UPE enabled
- SEPA within UPE enabled
- Subscriptions plugin installed
- A subscription with a free trial created

Testing:
- As the customer add the free trial subscription product to the cart.
- Go to checkout with only the free trial subscription product and confirm that UPE is displayed with SEPA as an option.
- Checkout using the following IBAN to simulate an immediate success: `AT611904300234573201`
- Confirm that the order was received successfully
- As the merchant, view the subscription and confirm it was created successfully.
- As the merchant, view the subscription and select to process the renewal under Subscription actions.
- Watch for the charge.succeeded webhook to fire and then confirm that the renewal order is marked as `processing`.

-------------------
